File: BOJ/1167.cpp
#include <climits>
#include <iostream>
#include <vector>

using namespace std;

struct Edge {
	int to;
	int distance;
};

struct Node {
	bool all_children_visited;
	int max_distance_to_leaf;
	vector<Edge> edges;
};

int find_maximum_diameter(Node &node, vector<Node> &nodes) {
	int first_max_distance_to_leaf = 0;
	int second_max_distance_to_leaf = 0;
	for (Edge& edge : node.edges) {
		// update max_distance_to_leaf
		int distance_to_leaf = edge.distance + nodes[edge.to].max_distance_to_leaf;
		if (distance_to_leaf >= first_max_distance_to_leaf) {
			second_max_distance_to_leaf = first_max_distance_to_leaf;
			first_max_distance_to_leaf = distance_to_leaf;
		} else if (distance_to_leaf > second_max_distance_to_leaf) {
			second_max_distance_to_leaf = distance_to_leaf;
		}
	}
	node.max_distance_to_leaf = first_max_distance_to_leaf;
	return first_max_distance_to_leaf + second_max_distance_to_leaf;
}

int main(void) {
	int V;
	int max_diameter = 0;

	vector<Node> nodes;
	vector<int> visiting_stack;

	ios::sync_with_stdio(false);
	cin.tie(NULL);

	// init all nodes (index 0 is dummy)
	cin >> V;
	nodes.resize(V + 1);
	for (Node &node : nodes) {
		node.all_children_visited = false;
		node.max_distance_to_leaf = INT_MIN;
	}

	while (V--) {
		// get a node number
		int v;
		cin >> v;

		// set adjacent edges
		while (true) {
			int to, distance;
			cin >> to;
			if (to == -1) break;
			cin >> distance;
			nodes[v].edges.emplace_back(Edge{to, distance});
		}
	}

	// start from any node
	// regard node 1 as root
	visiting_stack.emplace_back(1);
	while (!visiting_stack.empty()) {
		Node& node = nodes[visiting_stack.back()];

		// check if children are visited
		if (!node.all_children_visited) {
			node.all_children_visited = true;
			// visit others first
			for (Edge& edge : node.edges) {
				if (!nodes[edge.to].all_children_visited) {
					visiting_stack.emplace_back(edge.to);
				}
			}
			continue;
		}

		// if all children are visited,
		// then visit this node
		visiting_stack.pop_back();

		// update maximum diameter
		int diameter = find_maximum_diameter(node, nodes);
		if (diameter > max_diameter) {
			max_diameter = diameter;
		}
	}

	// print the maximum diameter
	cout << max_diameter;

	return 0;
}

File: BOJ/1167.ts
import * as fs from 'fs';

const inputManager: Generator<string, any, unknown> = (function *() {
  const stdin = '/dev/stdin';
  const lines = fs.readFileSync(stdin).toString().split('\n');
  for (const line of lines) yield line;
  return null;
})();
const nextLine = (): string => inputManager.next().value;
const nextLineNumbers = () => nextLine().split(' ').map(Number);

type Node = {
  allChildrenVisited: boolean;
  maxDistanceToLeaf: number;
  edges: Edge[];
};

type Edge = {
  to: Node;
  distance: number;
};

// init all nodes (index 0 is dummy)
let [V] = nextLineNumbers();
const nodes: Array<Node> = new Array(V + 1).fill(null).map((_, i) => ({
  allChildrenVisited: false,
  maxDistanceToLeaf: -Infinity,
  edges: [],
}));

let maxDiameter = 0;

const findMaximumDistanceToLeaf = (node: Node) => {
  let firstMaxDistanceToLeaf = 0;
  let secondMaxDistanceToLeaf = 0;

  for (const edge of node.edges) {
    // update maxDistanceToLeaf
    const distanceToLeaf = edge.distance + edge.to.maxDistanceToLeaf;
    if (distanceToLeaf >= firstMaxDistanceToLeaf) {
      secondMaxDistanceToLeaf = firstMaxDistanceToLeaf;
      firstMaxDistanceToLeaf = distanceToLeaf;
    } else if (distanceToLeaf > secondMaxDistanceToLeaf) {
      secondMaxDistanceToLeaf = distanceToLeaf;
    }
  }

  // update maximum diameter
  const diameter = firstMaxDistanceToLeaf + secondMaxDistanceToLeaf;
  if (diameter > maxDiameter) {
    maxDiameter = diameter;
  }

  return firstMaxDistanceToLeaf;
}

while (V--) {
  const [nodeNumber, ...edgeData] = nextLineNumbers();

  // set adjacent edges
  const node = nodes[nodeNumber];
  const nEdges = (edgeData.length - 1) / 2; // remove the end signal number (-1)
  for (let i = 0; i < nEdges; i++) {
    const to = edgeData[i * 2];
    const distance = edgeData[i * 2 + 1];
    node.edges.push({ to: nodes[to], distance });
  }
}

// start from any node
// regard node 1 as root
const visitingStack = [nodes[1]];
while (visitingStack.length) {
  const node = visitingStack[visitingStack.length - 1];

  // check if children are visited
  if (!node.allChildrenVisited) {
    node.allChildrenVisited = true;
    // visit others first
    visitingStack.push(...node.edges.filter(edge => !edge.to.allChildrenVisited).map(({to}) => to));
    continue;
  }

  // if all children are visited,
  // then visit this node
  visitingStack.pop();
  node.maxDistanceToLeaf = findMaximumDistanceToLeaf(node);
}

// print the maximum diameter
console.log(maxDiameter);
